<launch>

  <!-- Arguments -->
  <arg name="move_forward_only" default="false"/>

  <arg name="with_camera" default="true"/>
  <arg name="localization" default="false"/>
  <arg name="database_path" default="~/.ros/rtabmap.db"/>
  <arg     if="$(arg localization)" name="rtabmap_args" default=""/>
  <arg unless="$(arg localization)" name="rtabmap_args" default="-d"/>

  <!-- Creating RGBD image for first camera -->
  <group ns="camera1">
    <node pkg="rtabmap_sync" type="rgbd_sync" name="rgbd_sync1" output="screen">
      <remap from="rgb/image"         to="/camera_head/color/image_raw"/>
      <remap from="depth/image"       to="/camera_head/aligned_depth_to_color/image_raw"/>
      <remap from="rgb/camera_info"   to="/camera_head/color/camera_info"/>
      <!-- <remap from="depth/camera_info" to="/camera_head/depth/camera_info"/> -->

      <remap from="rgbd_image"       to="rgbd_image1"/>
    </node>
  </group>

  <!-- Launching RTAB-Map -->
  <group ns="rtabmap">

    <!-- Odometry -->
    <node pkg="rtabmap_odom" type="rgbd_odometry" name="rgbd_odometry" output="screen">
      <param name="subscribe_rgb"       type="bool" value="false"/>
      <param name="subscribe_depth"     type="bool" value="false"/>
      <param name="subscribe_rgbd" type="bool"   value="true"/>
      <param name="rgbd_cameras"   type="int"    value="1"/>
      <param name="frame_id"       type="string" value="camera_head_link"/>
      <param name="publish_tf"     type="bool"   value="true"/>
      <param name="wait_imu_to_init"  type="bool"   value="true"/>
      <remap from="rgbd_image"     to="/camera1/rgbd_image1"/>
      <remap from="imu"            to="/imu/data"/>
    </node>

    <node name="rtabmap" pkg="rtabmap_slam" type="rtabmap" output="screen" args="$(arg rtabmap_args)">
      <param name="database_path"       type="string" value="$(arg database_path)"/>
      <param name="frame_id"            type="string" value="camera_head_link"/>
      <param name="subscribe_rgb"       type="bool" value="false"/>
      <param name="subscribe_depth"     type="bool" value="false"/>
      <param name="subscribe_rgbd"      type="bool" value="true"/>
      <param name="rgbd_cameras"        type="int" value="1"/>
      <param name="subscribe_scan"      type="bool"   value="false"/>
      <param name="approx_sync"         type="bool"   value="true"/>
      <param name="publish_tf"          type="bool"   value="true"/>
      <param name="rtabmapviz"          type="bool"   value="true"/>

      <!-- inputs -->
      <remap from="rgbd_image"     to="/camera1/rgbd_image1"/>
      <!-- <remap from="rgbd_images"     to="/rtabmap/rgbd_images"/> -->
      <param name="queue_size" type="int" value="10"/>

      <!-- output -->
      <remap from="grid_map" to="/map"/>

      <!-- RTAB-Map's parameters -->
      <param name="Reg/Strategy"                 type="string" value="1"/>
      <param name="Reg/Force3DoF"                type="string" value="true"/>
      <param name="GridGlobal/MinSize"           type="string" value="20"/>
      <param name="Grid/Sensor"                  type="string" value="1"/>
      <param name="Vis/EstimationType"           type="string" value="0"/>  
      <param name="Grid/FromDepth"               type="string" value="true"/>
      <param name="Grid/RangeMax"                type="string" value="3.5"/>
      <param name="Grid/VoxelSize"               type="string" value="1"/>

      <!-- localization mode -->
      <param     if="$(arg localization)" name="Mem/IncrementalMemory" type="string" value="false"/>
      <param unless="$(arg localization)" name="Mem/IncrementalMemory" type="string" value="true"/>
    </node>

  </group>

</launch>